### ISSUE TITLE
endless loop

### ISSUE THREAD

**Issue by wojtekjacko** [spring-cloud-stream-app-starters/mqtt#3]
version SCDF 1.2.1 release deployed on PCF
When creating streaming by the following example:
 mqtt | processor | mqtt

meaning using both source component (from outside) and sink component (pointed somewhere outside)
the stream itself is deployed correct

however when putting message into this stream - sink component (having all headers information passed from the beginning of the stream) is putting message back to the "source"

in this circumstances - message never leaves this flow and it keeps looping inside

as workaround - I used to break the stream into two pieces as:
mqtt | processor | rabbit
rabbit | mqtt

however this increase number of flows and components used

it would be great if sink can reach out to outside world directly as
mqtt | mqtt

**Comment by artembilan** [comment]
This has been fixed in the https://jira.spring.io/browse/INT-4138.

You can configure `spring.cloud.stream.rabbit.bindings.<channelName>.consumer.headerPatterns`:
http://docs.spring.io/spring-cloud-stream/docs/Chelsea.SR2/reference/htmlsingle/index.html#_rabbitmq_binder

> **headerPatterns**
>    Patterns for headers to be mapped from inbound messages.
    Default: ['*'] (all headers).

To avoid undesired `mqtt_` headers mapping:
http://docs.spring.io/spring-integration/reference/html/amqp.html#amqp-message-headers

Pay attention to the:

> Starting with version 4.3, patterns in the header mappings can be negated by preceding the pattern with `!` .

**Comment by artembilan** [comment]
Meanwhile we are upgrading SCSt to Spring Integration 5.0, we can provide an out-of-the-box fix in to the MQTT Source to remap `mqtt_` headers to the `mqtt_received` - exactly what we do now Spring Integration 5.0.

**Comment by sabbyanandan** [comment]
Hi, @wojtekjacko. Thanks for reporting this.

As @artembilan points out, it is addressed in Spring Integration 5.0 _(requires Spring Boot 2.0)_ and to consume it in MQTT-source; we will have to upgrade the app to Spring Boot 2.0. 

However, we will be releasing 1.3 version of the Apps first and 2.0 is planned for early next year.

Until then, you could try the following optimization.

```
mqtt | processor > to_a _rabbit_queue_name
to_a _rabbit_queue_name > mqtt
```

This eliminates the need for extra rabbit applications. You're directly dropping and consuming it from a rabbit queue instead. Would this help?

**Comment by sabbyanandan** [comment]
Closing it due to inactivity.
